Generator fix

Ghostvv opened the pull request:
**Proposed changes**:
- perform deduplication in parallel with states generation in featurizers
- change deduplication logic in generator
- optimize performance for largely checkpointed stories
- add new parameter `unique_last_num_states` instead of `max_number_of_trackers`

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog

Ghostvv commented:
`test_remote_example` fails due to augmentation, Memo policy learns that these features are ambiguous

tmbo commented:
I am not sure what you mean with the mem policy. Can you explain?

Ghostvv commented:
So, augmentation creates additional stories that concatenate the end and the beginning, so it creates additional examples for some `actions`. The problem is that if they contradict the examples from original stories, `memo` policy treats these `actions` as ambiguous.

tmbo commented on rasa_core/training/structures.py:
```diff
@@ -289,9 +303,60 @@ def with_cycles_removed(self):
+
+    @staticmethod
+    def _unused_checkpoints(story_steps, story_end_checkpoints):
+        """Find all end checkpoints."""
```
description doesn't really fit the functions name.

tmbo commented on rasa_core/training/structures.py:
```diff
@@ -289,9 +303,60 @@ def with_cycles_removed(self):
+            for start in step.start_checkpoints:
+                collected_start.add(start.name)
+            for end in step.end_checkpoints:
+                if end.name in story_end_checkpoints.keys():
```
`end_name = story_end_checkpoints.get(end.name, end.name)` instead of if else

tmbo commented on rasa_core/training/structures.py:
```diff
@@ -289,9 +303,60 @@ def with_cycles_removed(self):
+
+        for k, step in list(story_steps.items()):
+            # changed all ends
+            updated = step.create_copy(use_new_id=False)
```
why do we need a copy if we replace it anyways?

tmbo commented on rasa_core/training/structures.py:
```diff
@@ -274,24 +284,89 @@ def with_cycles_removed(self):
+
+        for k, step in story_steps.items():
+            # changed all ends
+            updated = step.create_copy(use_new_id=False)
```
why do we need to copy this if we replace it anways?

Ghostvv replied:
do not know, i just used the previous code here.

tmbo replied:
ok :+1:

Ghostvv replied:
I guess because this step might be linked to another place in the list?

tmbo replied:
let's just keep it this way.

tmbo commented on rasa_core/training/structures.py:
```diff
@@ -274,24 +284,89 @@ def with_cycles_removed(self):
+                    end_name = end.name
+                collected_end.add(end_name)
+
+        return collected_end.union(collected_start)
```
I don't see how this function computes unused checkpoints ❓

Ghostvv replied:
well, you changed it to `union` :stuck_out_tongue_winking_eye:, it should be `^`=`XOR`

tmbo replied:
very good point :D this makes a lot more sense, changed it to `symmetric_difference` which is XOR.

Ghostvv commented on rasa_core/agent.py:
```diff
@@ -176,7 +176,7 @@ def load_data(self,
                   remove_duplicates=True,  # type: bool
                   augmentation_factor=20,  # type: int
-                  max_number_of_trackers=2000,  # type: int
+                  max_number_of_trackers=2000,  # type: Optional[int]
```
we also need to add `unique_last_num_states` there, I guess

tmbo replied:
yes

Ghostvv commented on rasa_core/training/generator.py:
```diff
@@ -59,8 +60,12 @@ def __init__(
         self.hashed_featurizations = set()
+        # story_graph.visualize('before_cycles_removed.pdf')
         self.story_graph = story_graph.with_cycles_removed()
+        self.story_graph.visualize('after_cycles_removed.pdf')
```
though I would add this functionality, it really helps, maybe a flag?

tmbo replied:
yes sounds good, but make it generic like `--debug_plots` also, please remove the commented line above

Ghostvv commented on rasa_core/training/generator.py:
```diff
@@ -92,77 +92,146 @@ def generate(self):
                         ts = start.filter_trackers(active_trackers[start.name])
                         incoming_trackers.extend(ts)
                         used_checkpoints.add(start.name)
+                    elif start.name not in used_checkpoints:
```
make sense. I think it is from my previous idea on counting unused checkpoints. i'll remove and test

Ghostvv replied:
I remembered. It is for the second and later generation rounds. If this checkpoint was added to `used` and does not exist in `active_trackers`, it means it was processed in previous rounds, and should not be processed again, like `STORY_START` for example.

Ghostvv commented on rasa_core/training/generator.py:
```diff
@@ -266,15 +348,52 @@ def _remove_duplicate_trackers(self, trackers):
+                    last_states = states[-self.unique_last_num_states:]
+                    last_hashed = self._hash_states(last_states)
+
+                    if last_hashed not in step_hashed_featurizations:
```
but this is idea, no? we process further only the trackers that are different in `unique_last_num_states` and all that duplicate these last steps send to directly to featurization.

tmbo replied:
ok thought about it again, sounds good :+1:

Ghostvv commented on rasa_core/training/generator.py:
```diff
@@ -187,41 +256,51 @@ def _subsample_trackers(self, incoming_trackers):
-            glue_mapping[STORY_END] = STORY_START
+        for end in unused_checkpoints:
+            # process trackers ended with unused checkpoints further
+            next_active_trackers[end].extend(active_trackers.get(end, []))
```
because active trackers might not have this `end` checkpoint in them, for example if there are unused start checkpoints created by users

Ghostvv replied:
I'll rename `end` here to `start` because we actually store only start checkpoints

tmbo replied:
but there is nio issue if `active_trackers ` doesn't contain `end`, that is why we are using get and a default value of `[]`. I am talking about the extend, which lets me assume that there are already elements in the `next_actove_trackers` that we are appending to. But as said, I don't think there are every any elements in there that we need to append to.

Ghostvv replied:
ahh, i do not know, again I just copied what was before, I'll take a look

Ghostvv commented on rasa_core/training/generator.py:
```diff
@@ -266,15 +348,52 @@ def _remove_duplicate_trackers(self, trackers):
 
         for tracker in trackers:
             states = self.domain.states_for_tracker_history(tracker)
             hashed = hash(tuple((frozenset(s) for s in states)))
```
if it is not too much trouble, I guess let's add them, since we have them present there. Right now I think the value is 1 if it is present, except for slots probably.

tmbo commented:
Did you look into the memoization policy yet?

Ghostvv commented:
do you mean problems with memoization and augmented stories? Yes, turned out we updated trackers with ActionReverted without copying them before, so deleted ActionListen in the original actions, so next time we use this tracker for augmentation, we deleted one more action

Ghostvv commented on rasa_core/training/__init__.py:
```diff
@@ -34,11 +34,12 @@ def load_data(
+        unique_last_num_states=None,  # type: Optional[int]
         augmentation_factor=20,  # type: int
-        max_number_of_trackers=2000,  # type: int
+        max_number_of_trackers=None,  # deprecated
```
but generator emits this warning, isn't it enough?

Ghostvv replied:
since all of them still call the generator eventually

tmbo replied:
my bad, in the pr view the generator is collapsed by default meaning search doesn't work, thats why I didn't find the warning. so just ignore this one :+1:

tmbo commented on rasa_core/agent.py:
```diff
@@ -175,17 +175,49 @@ def toggle_memoization(
     def load_data(self,
                   resource_name,  # type: Text
                   remove_duplicates=True,  # type: bool
+                  unique_last_num_states='auto',  # type: Optional[int]
```
let's keep it simple and go for None, if really needed, we can still fix it later.

Ghostvv replied:
Agree, if really needed one can use `training.load_data(...)` directly
